iceoryx_hoofs/source/posix_wrapper/thread.cpp

```diff
@@ -16,38 +16,97 @@
+    }
 }
 
-ThreadName_t getThreadName(pthread_t thread) noexcept
+void Thread::setThreadName(const ThreadName_t& name) noexcept
+{
+    posixCall(iox_pthread_setname_np)(m_threadHandle, name.c_str())
+        .successReturnValue(0)
+        .evaluate()
+        .or_else([](auto& r) {
+            // String length limit is ensured through cxx::string
+            // ERANGE (string too long) intentionally not handled to avoid untestable and dead code
+            LogError() << "This should never happen! " << r.getHumanReadableErrnum();
+            cxx::Ensures(false && "internal logic error");
+            /// @todo thread specific comm file under /proc/self/task/[tid]/comm is read. Opening this file can fail
+            /// and errors possible for open(2) can be retrieved. Handle them here?
+        });
+}
+
+ThreadName_t Thread::getThreadName() noexcept
 {
     char tempName[MAX_THREAD_NAME_LENGTH + 1U];
 
-    posixCall(pthread_getname_np)(thread, tempName, MAX_THREAD_NAME_LENGTH + 1U)
+    posixCall(pthread_getname_np)(m_threadHandle, tempName, MAX_THREAD_NAME_LENGTH + 1U)
         .successReturnValue(0)
         .evaluate()
         .or_else([](auto& r) {
             // String length limit is ensured through MAX_THREAD_NAME_LENGTH
             // ERANGE (string too small) intentionally not handled to avoid untestable and dead code
-            std::cerr << "This should never happen! " << r.getHumanReadableErrnum() << std::endl;
+            LogError() << "This should never happen! " << r.getHumanReadableErrnum();
             cxx::Ensures(false && "internal logic error");
+            /// @todo thread specific comm file under /proc/self/task/[tid]/comm is read. Opening this file can fail
+            /// and errors possible for open(2) can be retrieved. Handle them here?
         });
 
     return ThreadName_t(cxx::TruncateToCapacity, tempName);
 }
 
+ThreadError Thread::errnoToEnum(const int errnoValue) noexcept
+{
+    switch (errnoValue)
+    {
+    case EAGAIN:
+        LogError() << "insufficient resources to create another thread";
+        return ThreadError::INSUFFICIENT_RESOURCES;
+    case EINVAL:
+        LogError() << "invalid attribute settings";
+        return ThreadError::INVALID_ATTRIBUTES;
+    case ENOMEM:
+        LogError() << "not enough memory to initialize the thread attributes object";
+        return ThreadError::INSUFFICIENT_MEMORY;
+    case EPERM:
+        LogError() << "no appropriate permission to set required scheduling policy or parameters";
+        return ThreadError::INSUFFICIENT_PERMISSIONS;
+    default:
+        LogError() << "an unexpected error occurred in thread - this should never happen! errno: "
+                   << strerror(errnoValue);
+        return ThreadError::UNDEFINED;
+    }
+}
+
+void* Thread::startRoutine(void* callable)
+{
+    // necessary because the callback signature for pthread_create is void*(void*)
+    callable_t f = *static_cast<callable_t*>(callable);
```

Comment:
I looked closer at how it is invoked in `ThreadBuilder::create` with `&uninitializedThread->m_callable` and this is ok. I was under the impression we want to somehow support `void*` arguments (as a native pthread) and this caused me to make some mistake but we do not as of know.

To summarize:
1) We type erase with `void*` as `pthread_create` requires this.
2) We always pass `&uninitializedThread->m_callable`, which is stored as part of the object and always a `cxx::function<void(void)>`.
3) We wrap this locally to `f`, also a `cxx::function<void(void)>` (here I am not entirely sure whether it is a copy or wrapping the callable, but I think it is the latter).
4) We execute `f`.

The local copy is unfortunate though (or double wrap, see 3)). If it is a double wrap it is even more loss of performance (indirections again). 

Can you explain why it is needed? We should not move the thread while it is created and I assume `pthread_create` is blocking until everything is copied to it. 

It would be a problem if the thread is moved later than we will have a problem if we do not create this copy since the running thread would access the now invalid callable we moved from. But it cannot be moved as of now, so this cannot be it...

Considering this, I think we would still have a race here if we move after the thread was started, even with the copy to f.

1) pthread invoked Thread::startRoutine
2) copy to callable to `f` is initated but not yet completed
3) thread wrapper is moved
4) source of move is destroyed
5) copy to f resumes from an the destroyed callable (undefined)

TLDR: 
It is not wrong but we may be able to optimize it if we store the thread state elsewhere. I also think we do not need the local copy if we cannot move the thread but I may have overlooked something critical (it is fairly complex). I will have to rethink it some more and require more detail.

Reply:
@MatthiasKillat Your summary is correct. The local copy is not needed anymore and I removed it.